File: 60.Tarjan'sSCCAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio      */
/* -----------------------------------------------------------------------  */

import java.util.*;

class Graph {
    int V; // Number of vertices
    List<Integer>[] adj; // Adjacency list

    Graph(int V) {
        this.V = V;
        adj = new ArrayList[V];
        for (int i = 0; i < V; i++) adj[i] = new ArrayList<>();
    }

    void addEdge(int u, int v) {
        adj[u].add(v); // Add directed edge u -> v
    }
}

class TarjanSCC {
    Graph graph;
    int V;
    int index = 0; // DFS discovery time
    int[] indices;
    int[] lowLink;
    boolean[] onStack;
    Deque<Integer> stack;
    List<List<Integer>> sccs;

    TarjanSCC(Graph g) {
        this.graph = g;
        this.V = g.V;
        indices = new int[V];
        Arrays.fill(indices, -1); // Unvisited nodes marked -1
        lowLink = new int[V];
        onStack = new boolean[V];
        stack = new ArrayDeque<>();
        sccs = new ArrayList<>();
    }

    void strongConnect(int v) {
        indices[v] = index;
        lowLink[v] = index;
        index++;
        stack.push(v);
        onStack[v] = true;

        // Explore neighbors
        for (int w : graph.adj[v]) {
            if (indices[w] == -1) { // Node not visited
                strongConnect(w);
                lowLink[v] = Math.min(lowLink[v], lowLink[w]);
            } else if (onStack[w]) { // Node in current stack
                lowLink[v] = Math.min(lowLink[v], indices[w]);
            }
        }

        // Found root of SCC
        if (lowLink[v] == indices[v]) {
            List<Integer> scc = new ArrayList<>();
            int w;
            do {
                w = stack.pop();
                onStack[w] = false;
                scc.add(w);
            } while (w != v);
            sccs.add(scc);
        }
    }

    List<List<Integer>> getSCCs() {
        for (int v = 0; v < V; v++) {
            if (indices[v] == -1) strongConnect(v);
        }
        return sccs;
    }
}

public class Main {
    public static void main(String[] args) {
        Graph g = new Graph(8);
        int[][] edges = {{0,1},{1,2},{2,0},{2,3},{3,4},{4,5},{5,3},{6,5},{6,7}};
        for (int[] e : edges) g.addEdge(e[0], e[1]);

        TarjanSCC tarjan = new TarjanSCC(g);
        List<List<Integer>> sccs = tarjan.getSCCs();
        System.out.println("Strongly Connected Components: " + sccs);
    }
}

File: 60.Tarjan'sSCCAlgo.py
# Tarjan’s Algorithm is used to find Strongly Connected Components (SCCs) in a directed graph.
# An SCC is a maximal set of vertices where every vertex is 
# reachable from every other vertex in the set.

# Explanation

# Tarjan’s algorithm is DFS-based.

# It maintains discovery times and a low-link value for each node:

# discovery_time[v]: Time when node v was first visited.

# low_link[v]: The smallest discovery time reachable from v, including itself.

# A stack keeps track of the nodes in the current DFS path.

# Whenever a node's low_link is equal to its discovery_time, we have found a root 
# of an SCC, and we pop all nodes from the stack until we reach this node.

# The algorithm efficiently finds all SCCs in O(V + E) time.

from collections import defaultdict

class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.graph = defaultdict(list)

    def add_edge(self, u, v):
        self.graph[u].append(v)
    
class TarjanSCC:
    def __init__(self, graph):
        self.graph = graph
        self.V = graph.V
        self.index = 0
        self.stack = []
        self.on_stack = [False] * self.V
        self.indices = [-1] * self.V
        self.low_link = [-1] * self.V
        self.sccs = []

    def strongconnect(self, v):
        self.indices[v] = self.index
        self.low_link[v] = self.index
        self.index += 1
        self.stack.append(v)
        self.on_stack[v] = True

        for w in self.graph.graph[v]:
            if self.indices[w] == -1:
                self.strongconnect(w)
                self.low_link[v] = min(self.low_link[v], self.low_link[w])
            elif self.on_stack[w]:
                self.low_link[v] = min(self.low_link[v], self.indices[w])

        if self.low_link[v] == self.indices[v]:
            scc = []
            while True:
                w = self.stack.pop()
                self.on_stack[w] = False
                scc.append(w)
                if w == v:
                    break
            self.sccs.append(scc)

    def get_sccs(self):
        for v in range(self.V):
            if self.indices[v] == -1:
                self.strongconnect(v)
        return self.sccs
    
g = Graph(8)
edges = [(0,1),(1,2),(2,0),(2,3),(3,4),(4,5),(5,3),(6,5),(6,7)]
for u,v in edges:
    g.add_edge(u,v)

tarjan = TarjanSCC(g)
sccs = tarjan.get_sccs()
print("Strongly Connected Components:", sccs)

         
      


        

